src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs

```diff
@@ -0,0 +1,203 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Buffers.Binary;
+using System.Numerics;
+using System.Runtime.InteropServices;
+
+namespace Nethermind.Serialization.FluentRlp;
+
+public delegate void RefRlpWriterAction<in TContext>(ref RlpWriter arg, TContext value)
+    where TContext : allows ref struct;
+
+public delegate void RefRlpWriterAction(ref RlpWriter arg);
+
+public ref struct RlpWriter
+{
+    private const bool LengthMode = false;
+    private const bool ContentMode = true;
+
+    private bool _mode;
+
+    public int Length { get; private set; }
+
+    private byte[] _buffer;
+    private int _position;
+
+    public static RlpWriter LengthWriter()
+    {
+        return new RlpWriter
+        {
+            _mode = LengthMode
+        };
+    }
+
+    public static RlpWriter ContentWriter(byte[] buffer)
+    {
+        return new RlpWriter
+        {
+            _mode = ContentMode,
+            _buffer = buffer
+        };
+    }
+
+    public void Write<T>(T value) where T : unmanaged, IBinaryInteger<T>
+    {
+        switch (_mode)
+        {
+            case LengthMode:
+                LengthWrite(value);
+                break;
+            case ContentMode:
+                ContentWrite(value);
+                break;
+        }
+    }
+
+    private unsafe void LengthWrite<T>(T value) where T : unmanaged, IBinaryInteger<T>
+    {
+        var size = sizeof(T);
+        Span<byte> bigEndian = stackalloc byte[size];
+        value.WriteBigEndian(bigEndian);
+        bigEndian = bigEndian.TrimStart((byte)0);
```

Comment:
`TrimStart` does not seem to be heavily optimized. There might be something better that we can use, specially considering that we're removing leading zeros.

Reply:
Looks like .net doesn't use SIMD for it!
We could write Vector based way to find start index.
@benaadams